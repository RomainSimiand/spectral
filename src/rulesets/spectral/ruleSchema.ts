import { anEnum, array, object, optional, primitive } from './helpers';
import { ISchema } from './types';

export const ruleSchema: ISchema = {
  type: optional(anEnum(['validation', 'style'])),
  summary: optional(primitive('string')),
  description: optional(primitive('string')),
  severity: optional(anEnum([0, 1, 2, 3])),
  tags: optional(array(primitive('string'))),
  enabled: optional(primitive('boolean')),
  given: primitive('string'),
  when: optional(
    object({
      field: primitive('string'),
      pattern: optional(primitive('string')),
    })
  ),
  then: {
    type: 'or',
    types: [
      object({
        field: optional(primitive('string')),
        function: primitive('string'),
        functionOptions: optional(primitive('any')),
      }),
      array(
        object({
          field: optional(primitive('string')),
          function: primitive('string'),
          functionOptions: optional(primitive('any')),
        })
      ),
    ],
  },
};
