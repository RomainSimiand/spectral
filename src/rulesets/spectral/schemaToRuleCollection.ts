import { DiagnosticSeverity } from '@stoplight/types';
import { RuleCollection, RuleFunction, RuleType } from '../../types';
import { ISchema } from './types';

export function noUndefRuleKey(field: string) {
  return `${field}:no-undef`;
}

export function enumRuleKey(field: string) {
  return `${field}:enum-value-match`;
}

export function schemaToRuleCollection(schema: ISchema): RuleCollection {
  const ruleset: RuleCollection = {};
  Object.entries(schema).forEach(([key, value]) => {
    if (!value.optional) {
      ruleset[noUndefRuleKey(key)] = noUndefRule(key);
    }
    switch (value.type) {
      case 'enum':
        ruleset[enumRuleKey(key)] = enumRule(key, value.values);
        break;
    }
  });
  return ruleset;
}

function enumRule(key: string, values: any[]) {
  return {
    severity: DiagnosticSeverity.Error,
    type: RuleType.VALIDATION,
    summary: `'${key}' must be one of '${values.join(', ')}'`,
    given: '$.rules[*]',
    then: {
      field: key,
      function: RuleFunction.ENUM,
      functionOptions: {
        values,
      },
    },
  };
}

function noUndefRule(key: string) {
  return {
    severity: DiagnosticSeverity.Error,
    type: RuleType.VALIDATION,
    summary: `'${key}' must be defined`,
    given: '$.rules[*]',
    then: {
      field: key,
      function: RuleFunction.TRUTHY,
    },
  };
}
