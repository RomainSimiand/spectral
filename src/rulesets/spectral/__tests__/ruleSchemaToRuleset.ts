import { ruleSchema } from '../ruleSchema';
import { schemaToRuleCollection } from '../schemaToRuleCollection';

describe('convert rule schema to ruleset', () => {
  it('should generate a collection of rules automatically', () => {
    expect(schemaToRuleCollection(ruleSchema)).toMatchSnapshot();
  });
});
