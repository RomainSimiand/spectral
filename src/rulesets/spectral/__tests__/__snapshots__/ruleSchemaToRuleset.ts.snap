// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert rule schema to ruleset should generate a collection of rules automatically 1`] = `
Object {
  "given:no-undef": Object {
    "given": "$.rules[*]",
    "severity": 0,
    "summary": "'given' must be defined",
    "then": Object {
      "field": "given",
      "function": "truthy",
    },
    "type": "validation",
  },
  "severity:enum-value-match": Object {
    "given": "$.rules[*]",
    "severity": 0,
    "summary": "'severity' must be one of '0, 1, 2, 3'",
    "then": Object {
      "field": "severity",
      "function": "enumeration",
      "functionOptions": Object {
        "values": Array [
          0,
          1,
          2,
          3,
        ],
      },
    },
    "type": "validation",
  },
  "then:no-undef": Object {
    "given": "$.rules[*]",
    "severity": 0,
    "summary": "'then' must be defined",
    "then": Object {
      "field": "then",
      "function": "truthy",
    },
    "type": "validation",
  },
  "type:enum-value-match": Object {
    "given": "$.rules[*]",
    "severity": 0,
    "summary": "'type' must be one of 'validation, style'",
    "then": Object {
      "field": "type",
      "function": "enumeration",
      "functionOptions": Object {
        "values": Array [
          "validation",
          "style",
        ],
      },
    },
    "type": "validation",
  },
}
`;
